Kernel/Devices/SB16.cpp

```diff
@@ -184,114 +225,156 @@ void SB16::set_irq_line(u8 irq_number)
 {
-    m_irq_queue.wait_forever("SB16");
-    disable_irq();
+    auto& priv = card_private(card);
+    size_t offset = ((is_write ? priv.periods_written : priv.periods_read) % card.dma_periods) * card.bytes_per_period;
+    dbgln("Physical dma address offset: {}", offset);
+    return card.dma_region->physical_page(0)->paddr().offset(offset);
 }
 
-KResultOr<size_t> SB16::write(FileDescription&, size_t, const UserOrKernelBuffer& data, size_t length)
+u8* SB16::playback_current_dma_period(Card& card, bool is_write)
 {
-    if (!m_dma_region) {
-        auto page = MM.allocate_supervisor_physical_page();
-        if (!page)
-            return ENOMEM;
-        auto vmobject = AnonymousVMObject::create_with_physical_page(*page);
-        m_dma_region = MM.allocate_kernel_region_with_vmobject(*vmobject, PAGE_SIZE, "SB16 DMA buffer", Region::Access::Write);
-        if (!m_dma_region)
-            return ENOMEM;
-    }
+    auto& priv = card_private(card);
+    size_t offset = ((is_write ? priv.periods_written : priv.periods_read) % card.dma_periods) * card.bytes_per_period;
+    return card.dma_region->vaddr().as_ptr() + offset;
+}
 
-#if SB16_DEBUG
-    klog() << "SB16: Writing buffer of " << length << " bytes";
-#endif
-    VERIFY(length <= PAGE_SIZE);
-    const int BLOCK_SIZE = 32 * 1024;
-    if (length > BLOCK_SIZE) {
-        return ENOSPC;
-    }
+void SB16::setup_dma(Card& card, u32 length)
+{
+    const auto addr = playback_current_dma_period_paddr(card, false).get();
+    const u8 channel = m_is16bit ? 5 : 1; // 16-bit samples use DMA channel 5 (on the master DMA controller)
 
-    u8 mode = (u8)SampleFormat::Signed | (u8)SampleFormat::Stereo;
+    // Disable the DMA channel
+    IO::out8(m_is16bit ? 0xd4 : 0x0a, 4 + (channel % 4));
 
-    const int sample_rate = 44100;
-    set_sample_rate(sample_rate);
-    if (!data.read(m_dma_region->vaddr().as_ptr(), length))
-        return EFAULT;
-    dma_start(length);
+    // Clear the byte pointer flip-flop
+    IO::out8(m_is16bit ? 0xd8 : 0x0c, 0);
 
-    // 16-bit single-cycle output.
-    // FIXME: Implement auto-initialized output.
-    u8 command = 0xb0;
+    // Write the DMA mode for the transfer
+    IO::out8(m_is16bit ? 0xd6 : 0x0b, (channel % 4) | 0x58);
 
-    u16 sample_count = length / sizeof(i16);
-    if (mode & (u8)SampleFormat::Stereo)
-        sample_count /= 2;
+    // Write the transfer length
+    auto transfer_length = (m_is16bit ? length / 2 : length) - 1;
+    IO::out8(m_is16bit ? 0xc6 : 0x03, (u8)transfer_length);
+    IO::out8(m_is16bit ? 0xc6 : 0x03, (u8)(transfer_length >> 8));
 
-    sample_count -= 1;
+    // Write the buffer addresss
+    auto offset = (m_is16bit ? addr / 2 : addr) % 65536;
+    IO::out8(m_is16bit ? 0x8b : 0x83, addr >> 16);
+    IO::out8(m_is16bit ? 0xc4 : 0x02, (u8)offset);
+    IO::out8(m_is16bit ? 0xc4 : 0x02, (u8)(offset >> 8));
 
-    cli();
-    enable_irq();
```

Comment:
Where are SB16 IRQ's supposed to be enabled? I would guess this is why the system locks up, if you're expecting an IRQ.

Reply:
Yeah that was a miss, fixed that.